sale_fixed_discount/models/sale_order_line.py

```diff
@@ -14,45 +15,43 @@ class SaleOrderLine(models.Model):
         help="Fixed amount discount.",
     )
 
-    @api.onchange("discount")
-    def _onchange_discount_percent(self):
-        # _onchange_discount method already exists in core,
-        # but discount is not in the onchange definition
-        if self.discount:
-            self.discount_fixed = 0.0
+    @api.constrains("discount_fixed", "discount")
+    def _check_discounts(self):
+        """Check that the fixed discount and the discount percentage are consistent."""
+        for line in self:
+            if line.discount_fixed and line.discount:
+                currency = line.currency_id
+                calculated_fixed_discount = float_round(
+                    line._get_discount_from_fixed_discount(),
+                    precision_rounding=currency.rounding,
+                )
 
-    @api.onchange("discount_fixed")
+                if calculated_fixed_discount != line.discount:
+                    raise ValidationError(
+                        _(
+                            "The fixed discount %(fixed)s does not match the calculated "
+                            "discount %(discount)s %%. Please correct one of the discounts."
+                        )
+                        % {
+                            "fixed": line.discount_fixed,
+                            "discount": line.discount,
+                        }
+                    )
```

Comment:
In v15, only discount or discount_fixed was allowed. In the calculation they where implemented additive (even though you could not enter both, and also the further tax calculation wouldn't work properly.)

In v16 they are treated as one and will be both be set. This leads also to a different visualization in the PDF documents, because now both (fixed and %) is added to the PDF:
![image](https://github.com/OCA/sale-workflow/assets/1799080/4d57ba9c-f5f1-4faa-bfb5-bdc67a586166)
I think the change is not perfect because of the PDF output. 

I am not sure if we should allow fixed + % in future, when we get the tax calculation to handle this correctly. In the v15 tax calculation this wasn't easy to implement, but with the v16 changes this might be possible.
But when we allow both, we need to decide if we use:
price = (unit_price - fixed) * discount
or
price = (unit_price * discount) - fixed

I prefer the first one.

Reply:
@CRogos Actually, I personally disagree. The fixed discount should only serve as a _convenience_ implementation. All actual discounts are applied through the `discount` field. This also greatly simplifies the modules code.
So the workflow will always be `price = unit_price * discount` and nothing more.
I don't see the added benefit of splitting those.

I agree that the discount might be misleading on the PDF though. Perhaps we should not add a separate column for this and just do something like:
`$ 100.00 (10.0%)` Or the other way around. We could add the currency to it as clarification?

The initial bug you mentioned in your video also needs to be addressed. I'll follow up on the issue this Friday.

Thanks for testing!

Reply:
@PieterPaulussen I also had a quick view in the new tax calculation and you are right that even in the new implementation we should not change `price = unit_price * discount` . I like your proposal regarding the PDF `$100.00 (10%)`.

I have one more issue, when we use discount_fixed and discount as synonym. When you enter the discount_fixed, the discount is updated. But when you enter a discount, the dicount_fixed field stays on 0.00. 
Shouldn't we make dicount_fixed as computed column which directly updates discount?